torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -308,23 +266,31 @@ PyObject* THPVariable_getitem(PyObject* self, PyObject* index) {
   HANDLE_TH_ERRORS
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
   OptionalDeviceGuard device_guard(device_of(self_));
+  bool is_tracing = torch::jit::tracer::isTracing();
 
   // handle simple types: integers, slices, ellipsis
-  if (index == Py_None) {
-    return wrap(self_.unsqueeze(0));
-  } else if (index == Py_Ellipsis) {
-    return wrap(at::alias(self_));
-  } else if (THPUtils_checkLong(index)) {
-    return wrap(applySelect(self_, 0, index));
-  } else if (PySlice_Check(index)) {
-    return wrap(applySlice(self_, 0, index, true));
+  if (is_tracing) {
+    if (THPUtils_checkLong(index) && THPVariable_Check(index)) {
+      recordSelectTrace(THPVariable_Unpack(index));
+    } else if (PySlice_Check(index)) {
+      Tensor start_tensor, stop_tensor, step_tensor;
+      extractTensorsFromSlice(index, start_tensor, stop_tensor, step_tensor);
+      recordSliceTrace(start_tensor, stop_tensor, step_tensor);
+    }
+  }
+  if (index == Py_None || index == Py_Ellipsis || THPUtils_checkLong(index) || PySlice_Check(index)) {
```

Comment:
Why is the condition here, and the condition below, different?

Reply:
They are also different in the original code:
get_item:
https://github.com/pytorch/pytorch/blob/8195961f20caa18863698d85410d0818bfd67463/torch/csrc/autograd/python_variable_indexing.cpp#L312-L321
set_item:
https://github.com/pytorch/pytorch/blob/8195961f20caa18863698d85410d0818bfd67463/torch/csrc/autograd/python_variable_indexing.cpp#L380-L397

I think the main reason is that in the `get_item` path we need to treat bools as ByteTensors and invoke the advanced indexing path, while in the `set_item` path we don't need to do it. I am asking Sam why we have this distinction (it was added in https://github.com/pytorch/pytorch/pull/3725).

Reply:
From Sam:
> The choice of what to handle in the if/else was mostly copied from NumPy:
>
> https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/mapping.c#L229-L238
>
> NumPy doesn't handle True/False there. I don't remember why we're handling True/False in the setter path, but it might have been due to the lack of proper zero dim tensors or the lack of proper boolean tensors at the time.
>
> It might be a useful exercise to remove the Py_True/False checks and see if the advanced indexing tests still pass. If they fail, then we might have some bugs lurking in the advanced indexing code

Sam suggested that we can try removing the Py_True/False checks in the setter path and let the advanced indexing code handle the booleans, but I think this is outside the scope of this PR (and also has the risk slowing down the Py_True/False setter path because now they have to go through advanced indexing).